[pydrake] WrapDeprecated seems to strip out references

# Problem

Consider the following code:

```c++
class FooSystem : public systems::LeafSystem<double> {
 public:
  FooSystem() : systems::LeafSystem<double>() {
    port1_ =
        this->DeclareAbstractOutputPort("port1", &FooSystem::Calc1).get_index();
    port2_ =
        this->DeclareAbstractOutputPort("port2", &FooSystem::Calc1).get_index();
  }
  const systems::OutputPort<double>& get_output_port1() const {
    return systems::System<double>::get_output_port(port1_);
  }
  const systems::OutputPort<double>& get_output_port2() const {
    return systems::System<double>::get_output_port(port2_);
  }

 private:
  void Calc1(const systems::Context<double>& context, double* output) const {
    *output = context.get_time() + 1.0;
  }
  void Calc2(const systems::Context<double>& context, double* output) const {
    *output = context.get_time() + 2.0;
  }
  int port1_{-1};
  int port2_{-1};
};

void BindStuff(py::module m) {
  using Class = FooSystem;
  py::class_<Class> cls(m, "FooSystem", "class doc");
  cls  // BR
      .def(py::init(), "Constructor")
      .def("get_output_port1", &Class::get_output_port1,
          py_rvp::reference_internal, "Port 1 doc")
      .def("get_output_port2", 
          WrapDeprecated("Deprecated 2038-01-01", &Class::get_output_port2),
          py_rvp::reference_internal, "Port 2 doc");
}
```

We define a simple system with two output ports (the functionality of the port is irrelevant). We bind the class and the port accessors in the very simple, direct way.

In the bindings, the two output ports were bound in the same way initially. However, the code above shows the effort to deprecate output port 2. The attempt uses `WrapDeprecated`. This produces the following error (file paths and line numbers are meaningless, but the final error, pointing to `WrapDeprecated` is what's important):

```
In file included from bindings/pydrake/geometry_py.cc:9:
In file included from bazel-out/k8-opt/bin/bindings/pydrake/common/_virtual_includes/cpp_template_pybind/drake/bindings/pydrake/common/cpp_template_pybind.h:12:
In file included from bazel-out/k8-opt/bin/bindings/pydrake/common/_virtual_includes/cpp_param_pybind/drake/bindings/pydrake/common/cpp_param_pybind.h:14:
In file included from bazel-out/k8-opt/bin/bindings/pydrake/common/_virtual_includes/wrap_pybind/drake/bindings/pydrake/common/wrap_pybind.h:13:
bazel-out/k8-opt/bin/bindings/pydrake/common/_virtual_includes/wrap_function/drake/bindings/pydrake/common/wrap_function.h:53:12: error: call to deleted constructor of 'drake::systems::OutputPort<double>'
    return (self->*method)(std::forward<Args>(args)...);
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
bazel-out/k8-opt/bin/bindings/pydrake/common/_virtual_includes/deprecation_pybind/drake/bindings/pydrake/common/deprecation_pybind.h:79:26: note: in instantiation of function template specialization 'drake::pydrake::internal::infer_function_info<const drake::systems::OutputPort<double> &, drake::pydrake::(anonymous namespace)::FooSystem>' requested here
      message, internal::infer_function_info(std::forward<Func>(func)));
                         ^
bindings/pydrake/geometry_py.cc:1009:13: note: in instantiation of function template specialization 'drake::pydrake::WrapDeprecated<const drake::systems::OutputPort<double> &(drake::pydrake::(anonymous namespace)::FooSystem::*)() const>' requested here
            WrapDeprecated("Deprecated 2038-01-01", &Class::get_output_port2),
            ^
bazel-out/k8-opt/bin/tools/install/libdrake/_virtual_includes/drake_shared_library/drake/systems/framework/output_port.h:68:35: note: 'OutputPort' has been explicitly marked deleted here
  DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(OutputPort)
```

Ultimately, it's complaining that it's trying to bind a function whose return value can't be copied or moved. Somehow, the process of wrapping the function has changed the signature.

--- a/bindings/pydrake/common/test/wrap_function_test.cc
+++ b/bindings/pydrake/common/test/wrap_function_test.cc
@@ -5,6 +5,8 @@
 
 #include <gtest/gtest.h>
 
+#include "drake/common/drake_copyable.h"
+
 // @note Some of these statements violate style guide. That is intended, as
 // this should test references for use in `pybind`.
 
@@ -392,6 +394,26 @@ GTEST_TEST(WrapFunction, ChangeCallbackOnly) {
       // Arguments.
       double*, CallbackWrapped, CallbackWrapped>;
   check_expected::run(wrapped);
+}
+
+// Testing for #15505.
+struct ArgNoCopyNoMove {
+  DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(ArgNoCopyNoMove)
+  ArgNoCopyNoMove() {}
+};
+
+class ArgNoCopyNoMoveContainer {
+ public:
+  const ArgNoCopyNoMove& get_const() const { return noncopyable_; }
+  const ArgNoCopyNoMove& get_mutable() { return noncopyable_; }
+
+ private:
+  ArgNoCopyNoMove noncopyable_;
+};
+
+GTEST_TEST(WrapFunction, InferForArgNoCopyNoMove) {
+  WrapIdentity(&ArgNoCopyNoMoveContainer::get_const);
+  WrapIdentity(&ArgNoCopyNoMoveContainer::get_mutable);
 }
 
 }  // namespace pydrake

--- a/bindings/pydrake/common/wrap_function.h
+++ b/bindings/pydrake/common/wrap_function.h
@@ -40,7 +40,7 @@ auto infer_function_info(Return (*func)(Args...)) {
 // Infers `function_info<>` from a mutable method pointer.
 template <typename Return, typename Class, typename... Args>
 auto infer_function_info(Return (Class::*method)(Args...)) {
-  auto func = [method](Class* self, Args... args) {
+  auto func = [method](Class* self, Args... args) -> Return {
     return (self->*method)(std::forward<Args>(args)...);
   };
   return make_function_info<Return, Class*, Args...>(func);
@@ -49,7 +49,7 @@ auto infer_function_info(Return (Class::*method)(Args...)) {
 // Infers `function_info<>` from a const method pointer.
 template <typename Return, typename Class, typename... Args>
 auto infer_function_info(Return (Class::*method)(Args...) const) {
-  auto func = [method](const Class* self, Args... args) {
+  auto func = [method](const Class* self, Args... args) -> Return {
     return (self->*method)(std::forward<Args>(args)...);
   };
   return make_function_info<Return, const Class*, Args...>(func);

